### PR TITLE
[i2c] Allow zero-length requests

### PR DESCRIPTION
**Pull request by vireshk**
Commands like SMBUS Quick don't require a buffer for the request and are
called as zero-length requests. The specification allows such requests
under the VIRTIO_I2C_F_ZERO_LENGTH_REQUEST feature, which is mandatory
to be implemented by the devices now.

Add support for zero-length requests.

The [spec](https://github.com/oasis-tcs/virtio-spec/issues/112) changes are already approved and [linux](https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=94e5d5b04f8056f8eec57f8b104ec76d3e945279) code is queued too for upcoming merge window.

Signed-off-by: Viresh Kumar <viresh.kumar@linaro.org>

### CODE REVIEW

File: coverage_config_x86_64.json
Comment (stsquad) [inline coverage_config_x86_64.json:3]: it's a shame coverage keeps going down.

File: src/i2c/src/vhu_i2c.rs
Comment (stsquad) [inline src/i2c/src/vhu_i2c.rs:84]: It's odd we have two styles of bit declarations, one where we expect an explicit shift and one where we do it in the constant.

File: src/i2c/src/vhu_i2c.rs
Comment (stsquad) [inline src/i2c/src/vhu_i2c.rs:214]: so what happens if we don't have 3 descriptors - is that an invalid and unexpected condition?

File: coverage_config_x86_64.json
Comment (vireshk) [reply coverage_config_x86_64.json:3]: It goes up as well sometimes :)

File: src/i2c/src/vhu_i2c.rs
Comment (vireshk) [reply src/i2c/src/vhu_i2c.rs:84]: I agree to this, but this looked sane because of the way they are getting used eventually. 

VIRTIO_I2C_F_ZERO_LENGTH_REQUEST is used in features() where we use more (already defined) flags which don't have the shift within them.

While it makes the code more readable to make VIRTIO_I2C_FLAGS_M_RD contain the shift as it is used directly in the code later.

File: src/i2c/src/vhu_i2c.rs
Comment (vireshk) [reply src/i2c/src/vhu_i2c.rs:212]: No, it is a valid zero-length operation then, just that we don't need to process the buffer in that case.

File: coverage_config_x86_64.json
Comment (andreeaflorescu) [reply coverage_config_x86_64.json:3]: It looks a bit like a trend at this point with the coverage going down. What do you think about adding some tests first, and only after merging these PRs? This would help with providing the right infrastructure for adding tests for added functionality as the functionality gets merged. This should also bring more confidence that the functionality is correct.

Comment (vireshk) [reply coverage_config_x86_64.json:3]: This pull request do contain new test for the newly added feature and this pull request in particular needs to be merged now since it doesn't work with the new kernel otherwise.

On testing front I do agree that we need to do more. As I understood until now the coverage would increase only once we add tests for the code that  isn't getting tested at all, mostly vhu_i2c.rs i.e..

I think @mathieupoirier also had a look at that stuff. It wasn't really straight forward to implement and would likely need modification to other crates as well to provide mock implementations. And that's why it had been lagging since so long, I tried to look into it multiple times but wasn't able to figure out the right way of doing it.

Having said that, I think we can add more tests for the rest of the code as the layout is much better to add proper tests. I will work on that.

Comment (mathieupoirier) [reply coverage_config_x86_64.json:3]: Providing an adequate test infrastructure for the vring manipulation in vhu_i2c.rs and vhu_rng.rs is something that needs to be done.  As stated in my conversation with @andreeaflorescu [here](https://github.com/rust-vmm/vhost-device/pull/29), I may have found a way forward.  I will invest time on implementing that solution next week.

File: coverage_config_x86_64.json
Comment (stsquad) [inline coverage_config_x86_64.json:5]: ok no reason to hold up this PR then.

**Review by stsquad — APPROVED** [review]

File: coverage_config_x86_64.json
Comment (vireshk) [reply coverage_config_x86_64.json:5]: I will push this after https://github.com/rust-vmm/vhost-device/pull/62 gets merged

**Comment by vireshk** [comment]
Lets merge this now @stsquad @mathieupoirier

**Review by stsquad — APPROVED** [review]

**Review by mathieupoirier — COMMENTED** [review]
Other than the minor comment related to line 1035 in i2c.rs, this looks good.

File: src/i2c/src/i2c.rs
Comment (vireshk) [reply src/i2c/src/i2c.rs:1046]: Fixed.

**Review by mathieupoirier — APPROVED** [review]
Approving this despite the lower test coverage since #63 takes care of bumping that to 74.1.

**Review by stsquad — APPROVED** [review]
LGTM! :thumbsup:

**Pull request merged**
